tls/s2n_resume.c

```diff
@@ -868,18 +872,21 @@ int s2n_decrypt_session_ticket(struct s2n_connection *conn, struct s2n_stuffer *
     POSIX_GUARD(s2n_stuffer_skip_write(&state_stuffer, state_blob_size));
     POSIX_GUARD_RESULT(s2n_deserialize_resumption_state(conn, &from->blob, &state_stuffer));
 
-    uint64_t now = 0;
-    POSIX_GUARD_RESULT(s2n_config_wall_clock(conn->config, &now));
+    if (s2n_connection_get_protocol_version(conn) >= S2N_TLS13) {
+        return S2N_SUCCESS;
+    }
 
-    /* If the key is in decrypt-only state, then a new key is assigned
-     * for the ticket.
+    /* A new key is assigned for the ticket if the key is in decrypt-only state, 
+     * a key in encrypt-decrypt state is available, and TLS 1.2 is being used.
      */
+    uint64_t now = 0;
+    POSIX_GUARD_RESULT(s2n_config_wall_clock(conn->config, &now));
     if (now >= key->intro_timestamp + conn->config->encrypt_decrypt_key_lifetime_in_nanos) {
-        /* Check if a key in encrypt-decrypt state is available */
-        if (s2n_config_is_encrypt_decrypt_key_available(conn->config) == 1) {
+        bool key_available = false;
+        POSIX_GUARD_RESULT(s2n_config_is_encrypt_decrypt_key_available(conn->config, &key_available));
```

Comment:
It's interesting that we have a specific method to check that the key is available, instead of just retrieving the key using the same method we would when we actually need to use an encryption key. Why can't we just try to retrieve the key?

Reply:
I may not be seeing the bigger picture but the actual key is not being used inside this function. I believe it changes the state of the connection to signal new key needs to be assigned.

Reply:
I walked jouho through my concerns offline. I was thinking that a more standard pattern would be to use the same method to check and to retrieve to reduce the chance that the two methods accidentally disagreed, but that we weren't using the standard pattern here for a good reason. Retrieving is slightly more expensive, and since a key can expire between calls to retrieve, retrieve might disagree with itself anyway. So I assumed that our code already handled the case where checking for a key succeeded but later trying to retrieve a key failed. But after some investigation, I might be wrong and we might need to handle that case properly: https://github.com/aws/s2n-tls/issues/4528